components/raftstore/src/store/peer.rs

```diff
@@ -3137,6 +3141,32 @@ where
                 };
             }
         }
+        let flags = WriteBatchFlags::from_bits_truncate(req.get_header().get_flags());
+        if flags.contains(WriteBatchFlags::STALE_READ) {
+            let read_ts =
+                decode_var_u64(&mut req.mut_header().take_flag_data().as_slice()).unwrap();
+            let safe_ts = self.safe_ts.load(Ordering::Relaxed);
```

Comment:
These should be put into coprocessor.

Reply:
I will extract all of them after this PR and https://github.com/tikv/tikv/pull/9385 are merged

Reply:
Then there should be a todo.